### PR TITLE
Fixes issue #3558 kubeconfig using server paths does not work with tap

### PR DESCRIPTION
**Pull request by pietervogelaar**
Signed-off-by: Pieter Vogelaar <pieter@pietervogelaar.nl>

@grampelberg This pull request fixes issue #3558.

### CODE REVIEW

File: pkg/tap/tap.go [hunk 1]
```diff
@@ -37,7 +38,7 @@ func Reader(k8sAPI *k8s.KubernetesAPI, req *pb.TapByResourceRequest, timeout tim
 	if err != nil {
 		return nil, nil, err
 	}
-	url.Path = protohttp.TapReqToURL(req)
+	url.Path = path.Join(url.Path, protohttp.TapReqToURL(req))
```
Comment (adleong) [inline pkg/tap/tap.go]: This looks good but I wonder if we're going to run into this again for any other requests we make directly to the Kubernetes API server in the future.  Do you think we could have a more general solution by adding a RoundTripper to the client in `k8sAPI.NewClient()` which prepends url path prefix to request paths?

Comment (grampelberg) [reply pkg/tap/tap.go]: That already works correctly. This specific instance doesn’t work because we’re building the URL specific to tap requests and replacing the already constructed one by `NewClient()` wholesale here.

Comment (adleong) [reply pkg/tap/tap.go]: ah, you're right!  thanks!

**Comment by ihcsim** [comment]
@pietervogelaar While testing this branch with my local Rancher v2.3.1. I am getting an error with `tap` with the following command: `linkerd tap -n <ns> deploy`. Notice that only the resource kind is provided. The error is coming from [here](https://github.com/linkerd/linkerd2/blob/015ea9e17aa5673f90b2fe3b397a46aff6e9eba6/pkg/tap/tap.go#L61). Can you see if it happens on your end too? I can also reproduce the same error on my `kind` cluster, with this branch.

Steps to repro:
```sh
 ⚡  bin/linkerd version                                                                             
Client version: git-a848bc04
Server version: git-a848bc04

⚡  bin/linkerd tap -n linkerd deploy/linkerd-controller                                          
req id=3:0 proxy=in  src=192.168.1.66:35676 dst=10.42.0.17:9996 tls=not_provided_by_remote :method=GET :authority=10.42.0.17:9996 :path=/ping
rsp id=3:0 proxy=in  src=192.168.1.66:35676 dst=10.42.0.17:9996 tls=not_provided_by_remote :status=200 latency=308µs
end id=3:0 proxy=in  src=192.168.1.66:35676 dst=10.42.0.17:9996 tls=not_provided_by_remote duration=20µs response-length=5B
req id=3:1 proxy=in  src=192.168.1.66:37542 dst=10.42.0.17:9995 tls=not_provided_by_remote :method=GET :authority=10.42.0.17:9995 :path=/ping
rsp id=3:1 proxy=in  src=192.168.1.66:37542 dst=10.42.0.17:9995 tls=not_provided_by_remote :status=200 latency=394µs
end id=3:1 proxy=in  src=192.168.1.66:37542 dst=10.42.0.17:9995 tls=not_provided_by_remote duration=21µs response-length=5B

⚡  bin/linkerd tap -n linkerd deploy                                                          
Error: HTTP error, status Code [404] (unexpected API response: {
  "paths": [
    "/apis",
    "/apis/tap.linkerd.io",
    "/apis/tap.linkerd.io/v1alpha1",
    "/healthz",
    "/healthz/log",
    "/healthz/ping",
    "/metrics",
    "/openapi/v2",
    "/version"
  ]
})
```

For reasons unknown to me, on Rancher, this works:
```
https://<rancher-server-ip>/k8s/clusters/c-b9hjr/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments/linkerd-controller/tap
```
But this doesn't:
```
https://<rancher-server-ip>/k8s/clusters/c-b9hjr/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments/tap
```

For posterity, this is what I have to do to get Rancher to run locally as Docker containers :sweat_smile: 

1. Make sure my local `/etc/resolv.conf` name server to a non-local one so that the [CoreDNS pod doesn't crash](https://coredns.io/plugins/loop/#troubleshooting) (because Docker DNS)
1. Set up the Rancher server URL as `my-private-ip.nip.io`
1. For `kubectl proxy` to work, we have to include the Rancher-generated k8s cluster ID in the URL. E.g. `curl -v localhost:8001/k8s/clusters/c-b9hjr/apis/tap.linkerd.io/v1alpha1`.

LMK if you see the same error.

**Comment by pietervogelaar** [comment]
@ihcsim Yes, I experience the very same. My Rancher version is v2.2.9.

If I do `docker logs kube-apiserver` on a Rancher managed Kubernetes node:
```
I1029 15:46:52.720678       1 controller.go:107] OpenAPI AggregationController: Processing item v1alpha1.tap.linkerd.io
E1029 15:46:53.396759       1 controller.go:114] loading OpenAPI spec for "v1alpha1.tap.linkerd.io" failed with: ERROR $root.paths./apis/tap.linkerd.io/v1alpha1.get is missing required property: responses
ERROR $root.paths./apis/tap.linkerd.io.get is missing required property: responses
I1029 15:46:53.397021       1 controller.go:127] OpenAPI AggregationController: action for item v1alpha1.tap.linkerd.io: Rate Limited Requeue.
```

It's not in the log directly when I execute the tap command, but it may be related.

**Comment by grampelberg** [comment]
@pietervogelaar are you still interested in getting this across the finish line?

**Comment by pietervogelaar** [comment]
@grampelberg Certainly! But I don't know how to solve what @ihcsim mentioned.

**Comment by grampelberg** [comment]
@ihcsim any ideas? Where's the error being thrown?

**Comment by ihcsim** [comment]
@grampelberg The error is returned [here](https://github.com/linkerd/linkerd2/blob/015ea9e17aa5673f90b2fe3b397a46aff6e9eba6/pkg/tap/tap.go#L61). I haven't had a chance to investigate this further; the API extension stuff is quite new to me. I will have some time next week, if @pietervogelaar doesn't get to it first.

**Comment by cpretzer** [comment]
@ihcsim @grampelberg, I _think_t that this error message is addressed here: https://github.com/linkerd/linkerd2/pull/3700

Going to run some tests and if it works, then I'll approve this PR

**Comment by cpretzer** [comment]
@ihcsim @pietervogelaar I've reproduced this using `kind`, so it's not specific to rancher.

I'm going to keep debugging and will update this issue with more info when I have it.

Here's the output from both commands:

Successful when deployment specified:
```
bin/go-run cli tap -n linkerd deploy/linkerd-controller --verbose
DEBU[0000] Response from [https://127.0.0.1:57482/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments/linkerd-controller/tap] had headers: map[Content-Type:[application/octet-stream] Date:[Wed, 05 Feb 2020 05:50:20 GMT]] 
DEBU[0000] Waiting for data...                          
req id=0:0 proxy=in  src=10.244.0.8:41050 dst=10.244.0.5:9995 tls=not_provided_by_remote :method=GET :authority=10.244.0.5:9995 :path=/metrics
DEBU[0000] Waiting for data...                          
rsp id=0:0 proxy=in  src=10.244.0.8:41050 dst=10.244.0.5:9995 tls=not_provided_by_remote :status=200 latency=2805µs
```

404 error when no deployment is specified
```
bin/go-run cli tap -n linkerd deploy --verbose
DEBU[0000] Response from [https://127.0.0.1:57482/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments/tap] had headers: map[Content-Length:[205] Content-Type:[application/json] Date:[Wed, 05 Feb 2020 05:50:32 GMT]] 
Error: HTTP error, status Code [404] (unexpected API response: {
  "paths": [
    "/apis",
    "/apis/tap.linkerd.io",
    "/apis/tap.linkerd.io/v1alpha1",
    "/healthz",
    "/healthz/log",
    "/healthz/ping",
    "/metrics",
    "/openapi/v2",
    "/version"
  ]
})

```

File: pkg/tap/tap.go (hunk 1 above)
Comment (cpretzer) [reply pkg/tap/tap.go]: I figured out the cause for the error, and it is not unique to rancher.

According to the `url.Join` [docs](https://golang.org/pkg/path/#Join):

```
Join joins any number of path elements into a single path, adding a separating slash if necessary. The result is Cleaned; in particular, all empty strings are ignored.
```

This is relevant because the path for the kubernetes API uses two slashes in a row when an item is "namespaced". 

Here are examples of the commands that are executed and the k8s API URLs that are generated when
`url.Path = path.Join(url.Path, protohttp.TapReqToURL(req))` is replaced with 
`url.Path = fmt.Sprintf("%s%s",url.Path, protohttp.TapReqToURL(req))`:

command:
`
bin/go-run cli tap -n linkerd deploy/linkerd-controller --verbose
`
URL:
`
https://<rancher-url>/k8s/clusters/c-66q2l/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments/linkerd-controller/tap
`

command:
`
bin/go-run cli tap -n linkerd deploy --verbose
`
URL:
`https://<rancher-url>/k8s/clusters/c-66q2l/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments//tap
`

Notice the consecutive slashes `//` between `deployments` and `tap`


I get the same results when running against a single node `kind` cluster on my local machine.

To summarize, the *gotcha* here is that `join.Path` calls `join.Clean` which results in consecutive slashes `//` being _cleaned_ to `/`

**Comment by cpretzer** [comment]
@pietervogelaar are you still interested in working together to get this PR merged?

**Comment by cpretzer** [comment]
closing this in favor of #4151
